Rule module

diademiemi opened the pull request:
Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

This PR includes a new module, the `rule` module.

With this module you can list, create and delete rules in a ruleset.

Here's some example tasks that will:  
1. Create a new rule with the given options on site A  
2. List the rules in this ruleset on site A  
3. Take the first rule in this ruleset (The newly created rule) and create it on site B  
4. Delete the rule on site A  

The value of this module is that it allows importing and deleting rules with Ansible. With listing the rules and looping over it, you can sync rules between sites with Ansible.
It can also be used to list all the rules and create an export of this by exporting the output of this to a file with the Ansible copy module. Then you could use the slurp module to read this file again and import these rules.
For example we'll be using this on a schedule to keep a version history of rules and have an easy way to import rules from one site to another. Having rules saved as files and importable through this helps with recovery too.

<details><summary> rule.yml </summary>

```yml
- name: "Create a rule in checkgroup_parameters:memory_percentage_used."
  tribe29.checkmk.rule:
    server_url: "http://localhost/"
    site: "my_site"
    automation_user: "automation"
    automation_secret: "$SECRET"
    ruleset: "checkgroup_parameters:memory_percentage_used"
    rule:
        conditions: {
            "host_labels": [],
            "host_name": {
                "match_on": [
                    "server3"
                ],
                "operator": "one_of"
            },
            "host_tags": [],
            "service_labels": []
        }
        properties: {
            "comment": "Warning at 80%\nCritical at 90%\n",
            "description": "Allow higher memory usage",
            "disabled": false,
            "documentation_url": "https://github.com/tribe29/ansible-collection-tribe29.checkmk/blob/main/plugins/modules/rules.py"
        }
        value_raw: "{'levels': [80.0, 90.0]}"
    state: "present"

# List rules in ruleset.
- name: "List rules in ruleset."
  tribe29.checkmk.rule:
    server_url: "http://localhost/"
    site: "my_site"
    automation_user: "automation"
    automation_secret: "$SECRET"
    ruleset: "checkgroup_parameters:memory_percentage_used"
  register: rules

# Import first rule from this ruleset on a different site.
- name: "Import a rule on a different site."
  tribe29.checkmk.rule:
    server_url: "http://localhost/"
    site: "my_other_site"
    automation_user: "automation"
    automation_secret: "$SECRET"
    ruleset: "checkgroup_parameters:memory_percentage_used"
    rule: "{{ rules.response.value[0].extensions }}"
    state: "present"

# Delete first rule in this ruleset.
- name: "Delete a rule."
  tribe29.checkmk.rule:
    server_url: "http://localhost/"
    site: "my_site"
    automation_user: "automation"
    automation_secret: "$SECRET"
    ruleset: "checkgroup_parameters:memory_percentage_used"
    rule: "{{ rules.response.value[0].extensions }}"
    state: "absent"
```
</details>

diademiemi commented:
I'll make sure all tests run successfully once it's ready

diademiemi commented:
Got the module working and wrote some examples with it, let me know of any feedback you have!

diademiemi commented:
There's two bugs I noticed that I believe are on Checkmks end.
First, when exporting a rule, the value_raw is frequently messed up.
For example the raw_value will be `"{'levels': [80.0, 90.0]}"`. If I import this again, it won't work and Checkmk will complain that:
`Unable to read current options of this rule. Falling back to default values. When saving this rule now, your previous settings will be overwritten. Problem was: Levels: The datatype must be a tuple, but is list.`
When I export this same rule through the web UI, the output will be `"{'levels': (80.0, 90.0)}"`.
The difference here being the brackets. When exporting through the API the `value_raw` uses the incorrect brackets which makes importing break on most rules.

I could jankily replace all `[` and `]` with `(` and `)` which would probably fix a lot of rules but that's a very bad solution. I'm pretty sure this is an error on the Checkmk API. Could you let whoever maintains the API know that that export function results in exports that can't be imported? I doubt it'd be a very hard fix considering the value is exported correctly when done with the web UI, so the correct logic is already in the code.

The other bug I found is that the API call to list all rulesets results in an internal server error. I would love to use the `/domain-types/ruleset/collections/all` endpoint to be able to loop through all rules and export them all automatically. But currently this is the behavior:

<details> <summary> httpie output </summary>

```html
❯ http GET "$API_URL/domain-types/ruleset/collections/all" \
    "Authorization: Bearer $USER $PASSWORD" \
    "Accept: application/json"
HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 529
Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' ssh: rdp:; img-src 'self' data: https://*.tile.openstreetmap.org/ ; connect-src 'self' https://crash.checkmk.com/ https://license.checkmk.com/api/upload ; frame-ancestors 'self' ; base-uri 'self'; form-action 'self' javascript: 'unsafe-inline'; object-src 'self'; worker-src 'self' blob:
Content-Type: text/html; charset=iso-8859-1
Date: Mon, 18 Jul 2022 14:32:13 GMT
Permissions-Policy: accelerometer=(), autoplay=(), camera=(), encrypted-media=(), fullscreen=(), geolocation=(), gyroscope=(), magnetometer=(), microphone=(), midi=(), payment=(), picture-in-picture=(), usb=()
Referrer-Policy: origin-when-cross-origin
Server: Apache
X-Content-Type-Options: nosniff
X-Frame-Options: sameorigin
X-Permitted-Cross-Domain-Policies: none
X-XSS-Protection: 1; mode=block

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at 
 nobody@localhost to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
</body></html>
```

<details>

Let me know if I should also report this elsewhere.

kain88-de commented:
Which version of check mk are you testing against? If you can provide some example httpie scripts I can look at the errors. 

Regarding the internal server errors. The details are in the web.log file. You can find it with
```bash
sudo omd su <SITE>
cat var/log/web.log
```

diademiemi commented:
I've been testing on `2.1.0p4`.
The example command in the Redoc documentation causes this internal server error. 
```
http GET "$API_URL/domain-types/ruleset/collections/all" \
    "Authorization: Bearer $USER $PASSWORD" \
    "Accept: application/json"
```

The following error appears in `web.log`:
<details> <summary> web.log </summary>

```
2022-07-19 07:51:24,487 [40] [cmk.gui.wsgi.rest_api 10714] Unhandled exception (Crash-ID: 9636ceaa-0737-11ed-9242-525400e774c1)
Traceback (most recent call last):
  File "/omd/sites/site/lib/python3/cmk/gui/wsgi/applications/rest_api.py", line 483, in _wsgi_app
    return wsgi_app(environ, start_response)
  File "/omd/sites/site/lib/python3/cmk/gui/wsgi/applications/rest_api.py", line 240, in __call__
    wsgi_app = self.endpoint.wrapped(ParameterDict(path_args))
  File "/omd/sites/site/lib/python3/cmk/gui/plugins/openapi/restful_objects/decorators.py", line 838, in _wrapper
    response = func(param)
  File "/omd/sites/site/lib/python3/cmk/gui/plugins/openapi/restful_objects/decorators.py", line 702, in _validating_wrapper
    response = self.func(_params)
  File "/omd/sites/site/lib/python3/cmk/gui/plugins/openapi/endpoints/ruleset/__init__.py", line 57, in list_rulesets
    return serve_json(
  File "/omd/sites/site/lib/python3/cmk/gui/plugins/openapi/restful_objects/constructors.py", line 890, in serve_json
    response.set_data(json.dumps(data))
  File "/omd/sites/site/lib/python3.9/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/omd/sites/site/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/omd/sites/site/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/omd/sites/site/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type HTML is not JSON serializable
```
</details>

diademiemi commented:
As for the wrong export, when running this:
```
http GET "$API_URL/domain-types/rule/collections/all" \
    "Authorization: Bearer $USER $PASSWORD" \
    "Accept: application/json" \
    ruleset_name=='checkgroup_parameters:memory_percentage_used' 
```
I get the export of all rules in that folder. Under value[0].extensions.value_raw the value is `"{'levels': [90.0, 95.0]}"`

<details> <summary> Full output </summary>

```
❯ http GET "$API_URL/domain-types/rule/collections/all" \
    "Authorization: Bearer $USER $PASSWORD" \
    "Accept: application/json" \
    ruleset_name=='checkgroup_parameters:memory_percentage_used' \

HTTP/1.1 200 OK
Connection: Keep-Alive
Content-Length: 1110
Content-Security-Policy: default-src 'self' 'unsafe-inline' 'unsafe-eval' ssh: rdp:; img-src 'self' data: https://*.tile.openstreetmap.org/ ; connect-src 'self' https://crash.checkmk.com/ https://license.checkmk.com/api/upload ; frame-ancestors 'self' ; base-uri 'self'; form-action 'self' javascript: 'unsafe-inline'; object-src 'self'; worker-src 'self' blob:
Content-Type: application/json
Date: Tue, 19 Jul 2022 07:53:27 GMT
ETag: "597cd29ad05f230fffe28061c4446e2005cfc53d"
Keep-Alive: timeout=5, max=100
Permissions-Policy: accelerometer=(), autoplay=(), camera=(), encrypted-media=(), fullscreen=(), geolocation=(), gyroscope=(), magnetometer=(), microphone=(), midi=(), payment=(), picture-in-picture=(), usb=()
Referrer-Policy: origin-when-cross-origin
Server: Apache
X-Content-Type-Options: nosniff
X-Frame-Options: sameorigin
X-Permitted-Cross-Domain-Policies: none
X-XSS-Protection: 1; mode=block

{
    "domainType": "rule",
    "extensions": {
        "found_rules": 1
    },
    "id": "rule",
    "links": [
        {
            "domainType": "link",
            "href": "http://192.168.33.11/site/check_mk/api/1.0/domain-types/rule/collections/all",
            "method": "GET",
            "rel": "self",
            "type": "application/json"
        }
    ],
    "value": [
        {
            "domainType": "rule",
            "extensions": {
                "conditions": {
                    "host_labels": [],
                    "host_tags": [],
                    "service_labels": []
                },
                "folder": "/",
                "folder_index": 0,
                "properties": {
                    "description": "Allow higher memory usage",
                    "disabled": false
                },
                "ruleset": "checkgroup_parameters:memory_percentage_used",
                "value_raw": "{'levels': [90.0, 95.0]}"
            },
            "id": "e23a3aea-d9ff-4163-8e81-6c1f926087e0",
            "links": [
                {
                    "domainType": "link",
                    "href": "http://192.168.33.11/site/check_mk/api/1.0/objects/rule/e23a3aea-d9ff-4163-8e81-6c1f926087e0",
                    "method": "GET",
                    "rel": "self",
                    "type": "application/json"
                },
                {
                    "domainType": "link",
                    "href": "http://192.168.33.11/site/check_mk/api/1.0/objects/rule/e23a3aea-d9ff-4163-8e81-6c1f926087e0",
                    "method": "DELETE",
                    "rel": "urn:org.restfulobjects:rels/delete",
                    "type": "application/json"
                }
            ],
            "members": {},
            "title": "EXAMPLE Allow higher memory usage"
        }
    ]
}
```

When exporting it through the web UI, I get this
![image](https://user-images.githubusercontent.com/56408488/179697794-78eca52f-f149-460f-954d-ff39ec0f1581.png)

As you can see in the web UI parenthesis are used, this is correct and can be imported without issues. However the API response returns square brackets which when imported Checkmk will not accept. (The `Unable to read current options of this rule. Falling back to default values. When saving this rule now, your previous settings will be overwritten. Problem was: Levels: The datatype must be a tuple, but is list.`.)

lgetwan commented:
Thanks for this huge piece of code!
Hopefully, I'll be able to have a look at this in the next days.

diademiemi commented:
The internal server error in upon requesting all rulesets is fixed in 2.1.0p8! I didn't realise p4 wasn't the latest version and that's what I started on, lots of releases haha.

However exporting rules does still use the incorrect brackets.

lgetwan commented:
I opened an internal bug ticket for the problem with the wrong format in the response and assigned it to the REST-API team.

diademiemi commented:
Thank you!

robin-tribe29 commented:
For awareness: This PR stays in a waiting state, until the upstream issue is resolved.

@diademiemi if you feel like getting into it, you might want to create integration tests for this module. Take a look at the `tests` folder for reference.

diademiemi commented:
Sure. I was doing this on company time since we wanted this in our Ansible infrastructure before we migrate to 2.1, so this I'll be doing on my own time when I have the chance.

Also, the feature of this module having an output of all rules I think I should either remove or it should also be added to the hosts module. It's kind of inconsistent this way. And I had to do it for the hosts and just getting a list of rules/hosts is very easy just using the `ansible.builtin.uri` module, so removing it doesn't really take out any usecase.

lgetwan commented:
My internal ticket is not in progress on the developers' side, yet.

diademiemi commented:
Alright I see, hoping it will be resolved soon since this is a bug with a large impact on my workplaces migration to Checkmk 2.1

kain88-de commented:
I had a look at the bug today. This is unfortunately a more complicated problem. There is no good unambiguous mapping between wato files (python) and json. For example the rule `memory_percentage_used` expects a tuple for the levels, json has no way differentiation between a tuple and a list, therefore the braces in the GET return.

diademiemi commented:
I see. Could there be a way to request a specific format? The export which the web UI creates is correct, I'm assuming that's the JSON output since it's labeled as use for the API. 
![image](https://user-images.githubusercontent.com/56408488/184836555-4e5eb767-3c08-4827-b576-0b0330723773.png)


I do think it's a bit illogical that the output from the API uses the Python format, which doesn't work with the API to import it. If there could be an option to use the same format the web UI export uses that would be great

kain88-de commented:
The web UI does not create JSON output. If copies the raw value of the wato file in python format. The API returns always JSON, a sane choice. The problem is that wato rules are historically python files. I consider this a bigger problem and will push for a solution. But we first need to figure out internally how to best solve this.

diademiemi commented:
Ah right. So does the API export JSON, but when importing it uses the Python format? The web UI gets the value from the WATO file, which is the Python format. In that case the reverse of what I said I think would be a solution to at least allow this usecase for exporting/importing rules with the API. So having an option for the API also get the value from WATO, that way the output is the same as the input

Also worth noting that when using the API to import the JSON (outputted from the API), it gives an error that it can't be parsed when clicking on the rule in the web UI, but in the cases I've seen it does fill in the fields correctly. So it does seem that somehow it is able to parse it somewhat. But if the formats are different I doubt that works in all cases.

kain88-de commented:
The API is fixed and will be released in the next patch release. See the commit on the master branch for details. https://github.com/tribe29/checkmk/commit/c7c05900295b756ba22facf55e220ae424e325e0

diademiemi commented:
Awesome! Let's wait for this to release and then discuss further about this PR.

diademiemi commented:
I see 2.1.0p10 just released and includes this fix, awesome!
I'll test this soon and check if this allows us to proceed.

diademiemi commented:
As for my comment earlier

> Also, the feature of this module having an output of all rules I think I should either remove or it should also be added to the hosts module. It's kind of inconsistent this way. And I had to do it for the hosts and just getting a list of rules/hosts is very easy just using the ansible.builtin.uri module, so removing it doesn't really take out any usecase.

What are your thoughts on this? I think it could be powerful to return all rules so you can loop over them in Ansible, I feel like if we include this it should also be added to the other modules. And this is very niche.

But just doing the same with `ansible.builtin.uri` isn't complicated, so I could also remove it here and just use that for the very niche usecase. 

I'm leaning towards just removing it, documenting it would be more trouble than it's worth for the very specific function of looping over rules, when all it does is easily replaced by just a GET with `uri`.

diademiemi commented:
Just had the chance to test this on 2.1.0p10. Can confirm that for the purposes of this module the bug has been resolved.

Going to remove the response output. It's inconsistent with the rest of the plugins. If we want to include this I feel it should be included in the hosts and folder modules too, which should be a separate PR for a later date.

I'll try to create some integration tests for this module.

robin-tribe29 commented:
@lgetwan and @kain88-de: Can you two give your feedback, whether you think this is good to merge? I can not tell code-wise, but will try to do some testing later this week.

lgetwan commented on plugins/modules/rule.py:
```diff
@@ -0,0 +1,303 @@
+            "disabled": false,
+            "documentation_url": "https://github.com/tribe29/ansible-collection-tribe29.checkmk/blob/main/plugins/modules/rules.py"
+        }
+        value_raw: "{'levels': [80.0, 90.0]}"
```
During my tests, I had to use 
`value_raw: "{'levels': (80.0, 90.0)}"`
instead to make it work on a 2.1.0p11.

diademiemi replied:
Yes, true I'll fix that. These exports were made before 2.1.0p10, which fixed this export

lgetwan commented on plugins/modules/rule.py:
```diff
@@ -0,0 +1,303 @@
+            "disabled": false,
+            "documentation_url": "https://github.com/tribe29/ansible-collection-tribe29.checkmk/blob/main/plugins/modules/rules.py"
+        }
+        value_raw: "{'levels': [80.0, 90.0]}"
```
Here, also round brackets were needed.

lgetwan commented:
I added two (github-) comments in the code. They are just part of the Ansible documentation, though.

diademiemi commented on plugins/modules/rule.py:
```diff
@@ -0,0 +1,303 @@
+
+# If this is part of a collection, you need to use semantic versioning,
+# i.e. the version is of the form "2.5.0" and not "2.4".
+version_added: "0.8.0"
```
I updated it to 0.9.0

diademiemi commented on plugins/modules/rule.py:
```diff
@@ -0,0 +1,303 @@
+
+author:
+    - Robin Gierse (@robin-tribe29)
+    - Lars Getwan (@lgetwan)
```
Done!

robin-tribe29 commented:
I am currently trying to implement integration tests for this module, but I am struggling.
@diademiemi do you want to give it a shot yourself? You can use any existing integration test as a template.

diademiemi commented:
Yes, I'll try to make some time for that today

diademiemi commented:
Had a busy day, will begin on this now

diademiemi commented:
`ansible-test` didn't want to work on my machine so I'm pushing and hoping it miraculously works first try, if not I'll try getting it to work on my PC

diademiemi commented:
Looks like GitHub actions might not run the sanity tests for new modules?

I got ansible-test working on my machine so I'll test there..

diademiemi commented:
It didn't run with my integration test. I'm getting a 404 on the test, will look into it tomorrow

robin-tribe29 commented:
Sorry for the misunderstanding @diademiemi, the integration tests are in `tests/integration/targets`. There you can use one of the modules as a template. And we are talking about integration tests, not sanity tests. Those already apply to your module too.
But you will need local testing anyway, as GitHub will only run the tests from the `main` branch.

diademiemi commented:
Yeah I added some tests for it there. They run on my local machine now (With Podman, Docker won't work as always).

I'm still getting a 404 but I probably just messed up a variable somewhere. I've got an exam today so I'm a little busy but if I've got time I'll look into it.

lgetwan commented:
Good luck with your exam!

diademiemi commented:
Passed RHCE! 🎉

The 404 was caused by me giving a host, I didn't realise the host actually needed to exist for it to accept the input, I thought it was just metadata, kind of a nondescript error. Had some wrong loops from other tests too, will fix that and push

diademiemi commented:
```
TASK [rule : Testing.] ******************************************************************************************************************************************************************************************************************
included: /root/ansible_collections/tribe29/checkmk/tests/output/.tmp/integration/rule-j_0slctp-ÅÑŚÌβŁÈ/tests/integration/targets/rule/tasks/test.yml for testhost => (item={'version': '2.1.0p10', 'site': 'stable'})

TASK [rule : 2.1.0p10 - Create rule.] ***************************************************************************************************************************************************************************************************
changed: [testhost -> localhost] => {"changed": true, "msg": "Created rule"}

TASK [rule : 2.1.0p10 - Activate.] ******************************************************************************************************************************************************************************************************
changed: [testhost -> localhost] => {"changed": true, "http_code": 200, "msg": "Changes activated."}

TASK [rule : 2.1.0p10 - Create rule.] ***************************************************************************************************************************************************************************************************
ok: [testhost -> localhost] => {"changed": false, "msg": "Rule already exists"}

TASK [rule : 2.1.0p10 - Fail if changed.] ***********************************************************************************************************************************************************************************************
skipping: [testhost] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [rule : 2.1.0p10 - Delete rule.] ***************************************************************************************************************************************************************************************************
changed: [testhost -> localhost] => {"changed": true, "msg": "Deleted rule"}

TASK [rule : 2.1.0p10 - Activate.] ******************************************************************************************************************************************************************************************************
changed: [testhost -> localhost] => {"changed": true, "http_code": 200, "msg": "Changes activated."}

TASK [rule : 2.1.0p10 - Delete rule.] ***************************************************************************************************************************************************************************************************
ok: [testhost -> localhost] => {"changed": false, "msg": "Rule did not exist"}

TASK [rule : 2.1.0p10 - Fail if changed.] ***********************************************************************************************************************************************************************************************
skipping: [testhost] => {"changed": false, "skip_reason": "Conditional result was False"}

PLAY RECAP ******************************************************************************************************************************************************************************************************************************
testhost                   : ok=14   changed=7    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   

```

diademiemi commented:
Added some fail statements to test idem potency too

robin-tribe29 commented:
> Passed RHCE! tada

Awesome, congratulations! :tada: I can only offer `LPIC 1`, which leaves me a little envious. :wink: 

I'll take a look at your tests and report back later.

diademiemi commented:
Awesome! I'll review your comments in the PR for 0.9.0 to get that ready for release

kain88-de commented on README.md:
```diff
@@ -59,7 +59,7 @@ Name | Description | Tests
 [tribe29.checkmk.downtime](https://github.com/tribe29/ansible-collection-tribe29.checkmk/blob/main/plugins/modules/downtime.py)|Schedule downtimes on hosts and services.|[![Integration Tests for Downtime Module](https://github.com/tribe29/ansible-collection-tribe29.checkmk/actions/workflows/ans-int-test-downtime.yaml/badge.svg)](https://github.com/tribe29/ansible-collection-tribe29.checkmk/action ...
 [tribe29.checkmk.folder](https://github.com/tribe29/ansible-collection-tribe29.checkmk/blob/main/plugins/modules/folder.py)|Manage folders.|[![Integration Tests for Folder Module](https://github.com/tribe29/ansible-collection-tribe29.checkmk/actions/workflows/ans-int-test-folder.yaml/badge.svg)](https://github.com/tribe29/ansible-collection-tribe29.checkmk/actions/workflows/ans-int-test-folder.ya ...
-[tribe29.checkmk.host](https://github.com/tribe29/ansible-collection-tribe29.checkmk/blob/main/plugins/modules/host.py)|Manage hosts.|[![Integration Tests for Host Module](https://github.com/tribe29/ansible-collection-tribe29.checkmk/actions/workflows/ans-int-test-host.yaml/badge.svg)](https://github.com/tribe29/ansible-collection-tribe29.checkmk/actions/workflows/ans-int-test-host.yaml)
+[tribe29.checkmk.rule](https://github.com/tribe29/ansible-collection-tribe29.checkmk/blob/main/plugins/modules/rule.py)|Manage rules.|[![Integration Tests for Rule Module](https://github.com/tribe29/ansible-collection-tribe29.checkmk/actions/workflows/ans-int-test-rule.yaml/badge.svg)](https://github.com/tribe29/ansible-collection-tribe29.checkmk/actions/workflows/ans-int-test-rule.yaml)
```
The host module still exists. Can you add the line for the rule module instead of replacing `host` with `rule`.

diademiemi replied:
Yes, added it back

kain88-de reviewed (approved):
Only a tiny nitpick in the readme. Looks good. Thanks
